### ISSUE TITLE
3.4 Calico node inconsistent behaviour when generating bird.cfg when acting as route reflector

### ISSUE THREAD

**Issue by mrrandrade** [projectcalico/calico#2373]
We're trying to replace a 3.2 route reflector setup with a 3.4 cluster with calico-node as route reflectors.

We have followed [the guide](https://docs.projectcalico.org/v3.4/usage/configuration/bgp) but the result, without a single a change in configuration, that sometimes it generates the /etc/calico/confd/config/bird.cfg correctly, sometimes it generates with the default AS number instead. 

## Expected Behavior
 Every execution to generate the correct /etc/calico/confd/config/bird.cfg

## Current Behavior
 Inconsistence of generating /etc/calico/confd/config/bird.cfg.

## Possible Solution
Would like some help on pointing where to look for this one.

## Steps to Reproduce (for bugs)

We have followed the BGP configuration guide linked above.

1. We have set up the nodes with labels and routereflector cluster id:

```
calicoctl get nodes 10.79.63.19 -o yaml --export  
apiVersion: projectcalico.org/v3
kind: Node
metadata:
  creationTimestamp: null
  labels:
    route-reflector: "true"
  name: 10.79.63.19
spec:
  bgp:
    ipv4Address: 10.79.63.19/32
    routeReflectorClusterID: 224.0.0.1
apiVersion: projectcalico.org/v3
kind: Node
metadata:
  creationTimestamp: null
  labels:
    route-reflector: "true"
  name: 10.79.63.20
spec:
  bgp:
    ipv4Address: 10.79.63.20/32
    routeReflectorClusterID: 224.0.0.1
```

This is our bgpconfiguration object. Notice we do not use the default value (64512).

```
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  creationTimestamp: null
  name: default
spec:
  asNumber: 64513
  logSeverityScreen: Info
  nodeToNodeMeshEnabled: false
```

We have fully followed the guide and created all kinds of different types of configuration of bgppers in order to figure out what is wrong. This is our current configuration:

```
apiVersion: projectcalico.org/v3
kind: BGPPeer
metadata:
  creationTimestamp: null
  name: peer-to-rrs
spec:
  asNumber: 0
  nodeSelector: '!has(route-reflector)'
  peerIP: ""
  peerSelector: has(route-reflector)
```

(I noted that the documentation has  nodeSelector: !has(route-reflector) without quotes, and as is, it does not get recorded into the ETCD backend. But anyways, We've tried with it, without it, and we still get inconsistent behaviour).

As a result of this configuration, we expect all of our nodes to pair with both route reflectors.

MOST of the times, when we start calico-node on route reflector, the bird.cfg shows up with the peers set with the WRONG as number (64512, which is the default). 

There is no occurence of 64512 on the ETCD database, so I'm assuming confd is either setting default for some reason:

```
# cat /etc/calico/confd/config/bird.cfg
...
# Template for all BGP clients
template bgp bgp_template {
  debug { states };
  description "Connection to BGP peer";
  # this is the correct number, as configured in the bgpconfiguration object.
  local as 64513;
  multihop;

...

# For peer /host/10.79.63.20/peer_v4/10.79.63.11
protocol bgp Node_10_79_63_11 from bgp_template {
  # When it DOESN'T work, the as get set to 64512 somehow and I do not know where it comes from.
  neighbor 10.79.63.11 as 64512;
}
 
# For peer /host/10.79.63.20/peer_v4/10.79.63.12
protocol bgp Node_10_79_63_12 from bgp_template {
  # When it DOESN'T work, the as get set to 64512 somehow and I do not know where it comes from.
  neighbor 10.79.63.12 as 64512;
}

```

Sometimes, we start it and without **any manual intervention** it just plain works.

```
# cat /etc/calico/confd/config/bird.cfg
...
# Template for all BGP clients
template bgp bgp_template {
  debug { states };
  description "Connection to BGP peer";
  # this is the correct number, as configured in the bgpconfiguration object.
  local as 64513;
  multihop;

...

# For peer /host/10.79.63.20/peer_v4/10.79.63.11
protocol bgp Node_10_79_63_11 from bgp_template {
  neighbor 10.79.63.11 as 64513;
  rr client;
  rr cluster id 224.0.0.1;
}
 
# For peer /host/10.79.63.20/peer_v4/10.79.63.12
protocol bgp Node_10_79_63_12 from bgp_template {
  # When this works, this line shows the correct AS and the rr client and cluster id shows up.
  neighbor 10.79.63.12 as 64513;
  rr client;
  rr cluster id 224.0.0.1;
}

```

One thing we noticed is that if they started in the "not working" setup, and we recreate a bgppeer object (without any changes, just a basic --export | replace), it **always** works.

I'd guess there's some problem sometimes with initial calico-node configuration; when reading data to process confd templates, maybe the order or processing might mess some values. After the recreation of a bgppeer object, it resyncs and corrects itself.

Could really use a hand on figuring out where to look, because this inconsistent behaviour is killing me!

## Context
 Basic setup with 2 route reflectors integrating with a regular router to broadcast Calico networks.

## Your Environment

* Calico version: 3.4 running on ETCD backend (not kubernetes internal)
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes 1.12, but I think its not relevant.
* Operating System and version: CoreOS 1911.3.0
* Link to your project (optional):

**Comment by mrrandrade** [comment]
This is confd log for a successfull run:

```
2019-01-17 01:14:43.656 [INFO][43] config.go 105: Skipping confd config file.
2019-01-17 01:14:43.656 [INFO][43] run.go 17: Starting calico-confd
2019-01-17 01:14:43.677 [INFO][43] watchersyncer.go 89: Start called
2019-01-17 01:14:43.677 [INFO][43] client.go 183: CALICO_ADVERTISE_CLUSTER_IPS not specified, no cluster ips will be advertised
2019-01-17 01:14:43.677 [INFO][43] client.go 319: RouteGenerator has indicated it is in sync
2019-01-17 01:14:43.677 [INFO][43] watchersyncer.go 127: Sending status update Status=wait-for-ready
2019-01-17 01:14:43.677 [INFO][43] watchersyncer.go 147: Starting main event processing loop
2019-01-17 01:14:43.678 [INFO][43] watchercache.go 271: Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
2019-01-17 01:14:43.679 [INFO][43] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:43.679 [INFO][43] watchersyncer.go 127: Sending status update Status=resync
2019-01-17 01:14:43.680 [INFO][43] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
2019-01-17 01:14:43.680 [INFO][43] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:43.680 [INFO][43] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
2019-01-17 01:14:43.681 [INFO][43] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"
2019-01-17 01:14:43.681 [INFO][43] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:43.681 [INFO][43] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:43.681 [INFO][43] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
2019-01-17 01:14:43.681 [INFO][43] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:43.681 [INFO][43] watchersyncer.go 221: All watchers have sync'd data - sending data and final sync
2019-01-17 01:14:43.682 [INFO][43] watchersyncer.go 127: Sending status update Status=in-sync
2019-01-17 01:14:43.682 [INFO][43] client.go 316: Calico Syncer has indicated it is in sync
2019-01-17 01:14:43.683 [INFO][43] resource.go 220: Target config /etc/calico/confd/config/bird_ipam.cfg out of sync
2019-01-17 01:14:43.703 [INFO][43] resource.go 220: Target config /etc/calico/confd/config/bird.cfg out of sync
2019-01-17 01:14:43.704 [INFO][43] resource.go 220: Target config /etc/calico/confd/config/bird6_aggr.cfg out of sync
2019-01-17 01:14:43.705 [INFO][43] resource.go 220: Target config /etc/calico/confd/config/bird6_ipam.cfg out of sync
2019-01-17 01:14:43.708 [INFO][43] resource.go 220: Target config /etc/calico/confd/config/bird_aggr.cfg out of sync
2019-01-17 01:14:43.710 [INFO][43] resource.go 220: Target config /etc/calico/confd/config/bird6.cfg out of sync
2019-01-17 01:14:43.711 [INFO][43] resource.go 220: Target config /tmp/tunl-ip out of sync
2019-01-17 01:14:43.712 [INFO][43] resource.go 260: Target config /etc/calico/confd/config/bird6_ipam.cfg has been updated
2019-01-17 01:14:43.712 [INFO][43] resource.go 260: Target config /etc/calico/confd/config/bird6_aggr.cfg has been updated
2019-01-17 01:14:43.712 [ERROR][43] resource.go 288: Error from checkcmd: "bird: /etc/calico/confd/config/.bird.cfg603625759, line 2: Unable to open included file /etc/calico/confd/config/bird_aggr.cfg: No such file or directory\n"
2019-01-17 01:14:43.712 [INFO][43] resource.go 226: Check failed, but file does not yet exist - create anyway
2019-01-17 01:14:43.722 [INFO][43] resource.go 260: Target config /etc/calico/confd/config/bird_aggr.cfg has been updated
2019-01-17 01:14:43.722 [INFO][43] resource.go 260: Target config /etc/calico/confd/config/bird_ipam.cfg has been updated
2019-01-17 01:14:43.723 [INFO][43] resource.go 260: Target config /etc/calico/confd/config/bird.cfg has been updated
2019-01-17 01:14:43.733 [INFO][43] resource.go 260: Target config /etc/calico/confd/config/bird6.cfg has been updated
2019-01-17 01:14:43.789 [INFO][43] resource.go 260: Target config /tmp/tunl-ip has been updated

```

This one is a fail:
```
2019-01-17 01:14:29.667 [INFO][45] config.go 105: Skipping confd config file.
2019-01-17 01:14:29.667 [INFO][45] run.go 17: Starting calico-confd
2019-01-17 01:14:29.684 [INFO][45] watchersyncer.go 89: Start called
2019-01-17 01:14:29.684 [INFO][45] client.go 183: CALICO_ADVERTISE_CLUSTER_IPS not specified, no cluster ips will be advertised
2019-01-17 01:14:29.684 [INFO][45] client.go 319: RouteGenerator has indicated it is in sync
2019-01-17 01:14:29.684 [INFO][45] watchersyncer.go 127: Sending status update Status=wait-for-ready
2019-01-17 01:14:29.684 [INFO][45] watchersyncer.go 147: Starting main event processing loop
2019-01-17 01:14:29.686 [INFO][45] watchercache.go 271: Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
2019-01-17 01:14:29.686 [INFO][45] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:29.686 [INFO][45] watchersyncer.go 127: Sending status update Status=resync
2019-01-17 01:14:29.687 [INFO][45] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
2019-01-17 01:14:29.687 [INFO][45] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:29.687 [INFO][45] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"
2019-01-17 01:14:29.687 [INFO][45] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:29.689 [INFO][45] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
2019-01-17 01:14:29.690 [INFO][45] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
2019-01-17 01:14:29.690 [INFO][45] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:29.690 [INFO][45] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:29.690 [INFO][45] watchersyncer.go 221: All watchers have sync'd data - sending data and final sync
2019-01-17 01:14:29.691 [INFO][45] watchersyncer.go 127: Sending status update Status=in-sync
2019-01-17 01:14:29.691 [INFO][45] client.go 316: Calico Syncer has indicated it is in sync
2019-01-17 01:14:29.693 [INFO][45] resource.go 220: Target config /etc/calico/confd/config/bird_ipam.cfg out of sync
2019-01-17 01:14:29.694 [INFO][45] resource.go 220: Target config /tmp/tunl-ip out of sync
2019-01-17 01:14:29.698 [INFO][45] resource.go 220: Target config /etc/calico/confd/config/bird.cfg out of sync
2019-01-17 01:14:29.701 [INFO][45] resource.go 260: Target config /etc/calico/confd/config/bird_ipam.cfg has been updated
2019-01-17 01:14:29.703 [ERROR][45] resource.go 288: Error from checkcmd: "bird: /etc/calico/confd/config/.bird.cfg389327987, line 2: Unable to open included file /etc/calico/confd/config/bird_aggr.cfg: No such file or directory\n"
2019-01-17 01:14:29.703 [INFO][45] resource.go 226: Check failed, but file does not yet exist - create anyway
2019-01-17 01:14:29.713 [INFO][45] resource.go 220: Target config /etc/calico/confd/config/bird_aggr.cfg out of sync
2019-01-17 01:14:29.715 [INFO][45] resource.go 220: Target config /etc/calico/confd/config/bird6_ipam.cfg out of sync
2019-01-17 01:14:29.717 [INFO][45] resource.go 260: Target config /etc/calico/confd/config/bird.cfg has been updated
2019-01-17 01:14:29.717 [INFO][45] resource.go 220: Target config /etc/calico/confd/config/bird6_aggr.cfg out of sync
2019-01-17 01:14:29.728 [INFO][45] resource.go 220: Target config /etc/calico/confd/config/bird6.cfg out of sync
2019-01-17 01:14:29.731 [INFO][45] resource.go 260: Target config /etc/calico/confd/config/bird6_ipam.cfg has been updated
2019-01-17 01:14:29.732 [INFO][45] resource.go 260: Target config /etc/calico/confd/config/bird_aggr.cfg has been updated
2019-01-17 01:14:29.734 [INFO][45] resource.go 260: Target config /etc/calico/confd/config/bird6_aggr.cfg has been updated
2019-01-17 01:14:29.738 [INFO][45] resource.go 260: Target config /etc/calico/confd/config/bird6.cfg has been updated
2019-01-17 01:14:29.782 [INFO][45] resource.go 260: Target config /tmp/tunl-ip has been updated

```

This one is also a fail:

```
2019-01-17 01:14:20.981 [INFO][42] config.go 105: Skipping confd config file.
2019-01-17 01:14:20.982 [INFO][42] run.go 17: Starting calico-confd
2019-01-17 01:14:21.003 [INFO][42] watchersyncer.go 89: Start called
2019-01-17 01:14:21.003 [INFO][42] client.go 183: CALICO_ADVERTISE_CLUSTER_IPS not specified, no cluster ips will be advertised
2019-01-17 01:14:21.003 [INFO][42] client.go 319: RouteGenerator has indicated it is in sync
2019-01-17 01:14:21.006 [INFO][42] watchersyncer.go 127: Sending status update Status=wait-for-ready
2019-01-17 01:14:21.006 [INFO][42] watchersyncer.go 147: Starting main event processing loop
2019-01-17 01:14:21.009 [INFO][42] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
2019-01-17 01:14:21.009 [INFO][42] watchersyncer.go 127: Sending status update Status=resync
2019-01-17 01:14:21.009 [INFO][42] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:21.010 [INFO][42] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
2019-01-17 01:14:21.010 [INFO][42] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
2019-01-17 01:14:21.010 [INFO][42] watchercache.go 271: Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
2019-01-17 01:14:21.010 [INFO][42] watchercache.go 271: Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"
2019-01-17 01:14:21.010 [INFO][42] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:21.010 [INFO][42] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:21.010 [INFO][42] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:21.010 [INFO][42] watchersyncer.go 209: Received InSync event from one of the watcher caches
2019-01-17 01:14:21.010 [INFO][42] watchersyncer.go 221: All watchers have sync'd data - sending data and final sync
2019-01-17 01:14:21.011 [INFO][42] watchersyncer.go 127: Sending status update Status=in-sync
2019-01-17 01:14:21.011 [INFO][42] client.go 316: Calico Syncer has indicated it is in sync
2019-01-17 01:14:21.012 [INFO][42] resource.go 220: Target config /tmp/tunl-ip out of sync
2019-01-17 01:14:21.016 [INFO][42] resource.go 220: Target config /etc/calico/confd/config/bird.cfg out of sync
2019-01-17 01:14:21.020 [INFO][42] resource.go 220: Target config /etc/calico/confd/config/bird_ipam.cfg out of sync
2019-01-17 01:14:21.020 [INFO][42] resource.go 220: Target config /etc/calico/confd/config/bird6_ipam.cfg out of sync
2019-01-17 01:14:21.021 [INFO][42] resource.go 220: Target config /etc/calico/confd/config/bird_aggr.cfg out of sync
2019-01-17 01:14:21.022 [INFO][42] resource.go 220: Target config /etc/calico/confd/config/bird6.cfg out of sync
2019-01-17 01:14:21.022 [INFO][42] resource.go 220: Target config /etc/calico/confd/config/bird6_aggr.cfg out of sync
2019-01-17 01:14:21.026 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird6_ipam.cfg has been updated
2019-01-17 01:14:21.027 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird_aggr.cfg has been updated
2019-01-17 01:14:21.038 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird.cfg has been updated
2019-01-17 01:14:21.038 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird_ipam.cfg has been updated
2019-01-17 01:14:21.040 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird6.cfg has been updated
2019-01-17 01:14:21.046 [INFO][42] resource.go 260: Target config /etc/calico/confd/config/bird6_aggr.cfg has been updated
2019-01-17 01:14:21.098 [INFO][42] resource.go 260: Target config /tmp/tunl-ip has been updated

```

**Comment by mrrandrade** [comment]
After analyzing confd logs, i noticed, on **two successfull** executions, that this was the order of reading data:

```
Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"
```

```
Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
```

This is when it doesn't work:

```
Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
```
```
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/ippools"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/nodes"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgpconfigurations"
Sending synced update ListRoot="/calico/ipam/v2/host/10.79.63.20"
Sending synced update ListRoot="/calico/resources/v3/projectcalico.org/bgppeers"

```

**Comment by fasaxc** [comment]
@mrrandrade Thanks for the very detailed report; it's a great help to tracking down this sort of issue.

**Comment by mrrandrade** [comment]
> @mrrandrade Thanks for the very detailed report; it's a great help to tracking down this sort of issue.

Thanks for the attention.

I'd like to complement this issue with the fact that I assumed the problem would only occur on my route reflectors.

Today I restarted my kubernetes's masters (I have two of them, paired with two route reflectors). One of my masters came up okay with my cuystom AS Number; the other master got messed up with the default 64512 AS and couldn't pair with my route reflectors:

```
# ------------- Node-specific peers -------------
# For peer /host/10.79.63.11/peer_v4/10.79.63.19
protocol bgp Node_10_79_63_19 from bgp_template {
  neighbor 10.79.63.19 as 64512;
}
# For peer /host/10.79.63.11/peer_v4/10.79.63.20
protocol bgp Node_10_79_63_20 from bgp_template {
  neighbor 10.79.63.20 as 64512;
}
```

```
./calicoctl node status
Calico process is running.

IPv4 BGP status
+--------------+---------------+-------+----------+-----------------------------+
| PEER ADDRESS |   PEER TYPE   | STATE |  SINCE   |            INFO             |
+--------------+---------------+-------+----------+-----------------------------+
| 10.79.63.19  | node specific | start | 20:54:45 | Idle BGP Error: Bad peer AS |
| 10.79.63.20  | node specific | start | 20:54:50 | Idle BGP Error: Bad peer AS |
+--------------+---------------+-------+----------+-----------------------------+

IPv6 BGP status
No IPv6 peers found.
```

So the bug might affect any member of a calico network that has a custom AS number.

**Comment by briansan** [comment]
Hi @mrrandrade, I'm having trouble reproducing your issue. Here's what I've tried, please let me know if something about my procedure differs from yours.
1. step up cluster and install calico
2. label nodes as route-reflector
3. apply the bgp configuration resource with ASN: 64513
4. apply the bgp peer resources to hook up the route reflector

but after having done all that, I see the calico/node instance on the route reflector nodes having a bird config with ASN set to 64513.

**Comment by caseydavenport** [comment]
I think I've reproduced and fixed this here: https://github.com/projectcalico/confd/pull/202

**Comment by mrrandrade** [comment]
> Hi @mrrandrade, I'm having trouble reproducing your issue. Here's what I've tried, please let me know if something about my procedure differs from yours.

Have you restarted calico-node after you created bgpeer/bgpconfig? It only happens on start time to me. So it makes random reboots on servers particularly dangerous.

It's also not 100% of the time; yesterday I had it happen on regular nodes as well, not only route reflectors; that didn't happen before. I restarted the service three times (like beetlejuice :-P) and it worked. On another broken node, I just recreate one bgppeer and that triggers the update, it seems the update always fix things broken.

As a third note, are you using ETCD or kubernetes datastore? We have a cluster lab with kubernetes datastore and we didn't manage to get it to broken state, although I didn't try enough.

> I think I've reproduced and fixed this here: projectcalico/confd#202

That's great news! I think I'll take a sabbatical year to actually become a coder so I can get useful instead of just posting issues! Makes me sad.

**Comment by ianjuma** [comment]
@caseydavenport Is this available on 3.5? Experiencing the same issue. And re-applying the bgppeer always resolves it for me too. Experienced one on reboot today, with 3.5.0.

```
calicoctl node status
Calico process is running.

IPv4 BGP status
+--------------+---------------+-------+----------+-----------------------------+
| PEER ADDRESS |   PEER TYPE   | STATE |  SINCE   |            INFO             |
+--------------+---------------+-------+----------+-----------------------------+
| 10.4.0.105   | node specific | start | 18:23:38 | Idle BGP Error: Bad peer AS |
+--------------+---------------+-------+----------+-----------------------------+

IPv6 BGP status
No IPv6 peers found.

[root@worker ~]# calicoctl node status
Calico process is running.

IPv4 BGP status
+--------------+---------------+-------+----------+-------------------------------+
| PEER ADDRESS |   PEER TYPE   | STATE |  SINCE   |             INFO              |
+--------------+---------------+-------+----------+-------------------------------+
| 10.4.0.105   | node specific | start | 18:23:38 | Active BGP Error: Bad peer AS |
+--------------+---------------+-------+----------+-------------------------------+

IPv6 BGP status
No IPv6 peers found.

[root@worker ~]#
[root@worker ~]# calicoctl node status
Calico process is running.

IPv4 BGP status
+--------------+---------------+-------+----------+-------------------------------+
| PEER ADDRESS |   PEER TYPE   | STATE |  SINCE   |             INFO              |
+--------------+---------------+-------+----------+-------------------------------+
| 10.4.0.105   | node specific | start | 18:23:38 | Active BGP Error: Bad peer AS |
+--------------+---------------+-------+----------+-------------------------------+

IPv6 BGP status
No IPv6 peers found.

[root@worker1 ~]# calicoctl node status
Calico process is running.

IPv4 BGP status
+--------------+---------------+-------+----------+-------------+
| PEER ADDRESS |   PEER TYPE   | STATE |  SINCE   |    INFO     |
+--------------+---------------+-------+----------+-------------+
| 10.4.0.105   | node specific | up    | 18:25:01 | Established |
+--------------+---------------+-------+----------+-------------+

IPv6 BGP status
No IPv6 peers found.
```

Using the default ASN - upon a few restarts - it works fine, sometimes. But re-applying the config always seems to work.
